### PR TITLE
vc: Version support check is ineffective in createSandbox

### LINKED ISSUE #2311: Version support check is ineffective in createSandbox
# Description of problem

In cloudHypervisor#createSandbox, the check against major and minor supported versions cannot handle either version being 0.

# Expected result

The check should cover either major or minor version being 0.

# Actual result

### PR DESCRIPTION
**Pull request by tedyu**
Fixes #2311

If major version matches max supported major, we continue comparing the minor version.

Signed-off-by: Ted Yu <yuzhihong@gmail.com>

The error along with the original check was actually negated.
Since supportedMajorVersion is 0, the condition was never hit. But the new test exposes the bug.

### CODE REVIEW

File: virtcontainers/clh.go [hunk 1]
```diff
@@ -138,7 +138,7 @@ func (clh *cloudHypervisor) createSandbox(ctx context.Context, id string, networ
 			return err
 		}
 
-		if clh.version.Major < supportedMajorVersion && clh.version.Minor < supportedMinorVersion {
+		if clh.version.Major < supportedMajorVersion || (clh.version.Major == supportedMajorVersion && clh.version.Minor < supportedMinorVersion) {
```
Comment (jodh-intel) [inline virtcontainers/clh.go]: Ideally this change would be coupled with:

- A refactor so this logic is encapsulated into a function (something like `func (clh *cloudHypervisor) checkVersion()`).
- A table-based test exercising boundary cases for all the version variables used here (https://github.com/kata-containers/tests/blob/master/Unit-Test-Advice.md#table-driven-tests).

However, there was talk of this test being temporary (https://github.com/kata-containers/runtime/pull/2164/files#r345596109), so it's unclear to me how long this test will remain in the code? If it's "a while", I vote we add some tests :smile:

/cc @sameo, @ericooper, @jcvenegas.

**Comment by amshinde** [comment]
@tedyu looks good, but can you add some unit tests for this as well.

**Comment by tedyu** [comment]
cloudHypervisor#checkVersion accesses the version field.

Let me see if I can come up with some test.

File: virtcontainers/clh.go (hunk 1 above)
Comment (ericooper) [reply virtcontainers/clh.go]: I actually quite disagree. I in fact did not add this version check temporarily. the hypervisor binary is not part of the source tree so can be anything and it's cli syntax is still under constant revision which is fine. I did not add the builder pattern without a reason. In future versions of the hypervisor different builders can be instantiated based on that version to accommodate evolution of the cloud hypervisor cli. In production we see very often that systems break causing hours of troubleshooting because of these kind of implicit assumptions. This version check should help to avoid this. And although the current logic perhaps does not make perfectly sense it is there for future usage.

Comment (jodh-intel) [reply virtcontainers/clh.go]: Thanks for clarifying @ericooper. In which case, we definitely need unit tests for this imho.

Comment (tedyu) [reply virtcontainers/clh.go]: @jodh-intel 
I have added test for the version check.

Please take a look.

**Comment by tedyu** [comment]
@jodh-intel 
I have added message to the assertion.

File: virtcontainers/clh_test.go [hunk 2]
```diff
@@ -501,6 +502,53 @@ func TestClhCliVsock(t *testing.T) {
...
+		},
+	}
+	for _, tc := range testcases {
+		clh.version = CloudHypervisorVersion{
+			Major:    tc.major,
+			Minor:    tc.minor,
+			Revision: 0,
+		}
+		err := clh.checkVersion()
+		msg := fmt.Sprintf("%s, result: %v", tc.name, err)
```
Comment (jodh-intel) [inline virtcontainers/clh_test.go]: Please log `tc` as shown in the document. I would also add `chl.version` to `msg` personally. Think about how easy it is to debug a failing CI test running on a remote system where you only have access to the test log...

Comment (tedyu) [reply virtcontainers/clh_test.go]: I have updated the test.

**Comment by tedyu** [comment]
@jodh-intel 
Thanks for the detailed review.

If you can trigger ubuntu test, that would be nice.

**Comment by tedyu** [comment]
@ericooper 
Please review.

**Comment by tedyu** [comment]
@egernst 
Please trigger test again.

thanks

**Comment by tedyu** [comment]
The failure in debian test was not related to my PR:
```
17:39:31 --- FAIL: TestContainerListStatsWithIdSandboxIdFilter (64.59s)
17:39:31     container_stats_test.go:248: Create a pod config and run sandbox container
17:39:31     container_stats_test.go:256: Create container config and run containers in a pod
17:39:31     require.go:794: 
17:39:31         	Error Trace:	container_stats_test.go:272
17:39:31         	Error:      	Received unexpected error:
17:39:31         	            	rpc error: code = DeadlineExceeded desc = context deadline exceeded
17:39:31         	Test:       	TestContainerListStatsWithIdSandboxIdFilter
```

**Comment by tedyu** [comment]
From ubuntu run:
```
Skipping 2,272 KB..
```
It seems not all test log was preserved.

**Comment by grahamwhaley** [comment]
@tedyu - at the top of the console log at http://jenkins.katacontainers.io/job/kata-containers-runtime-ARM-18.04-PR/1304/console - you need to click on the 'Full Log' link to take you to http://jenkins.katacontainers.io/job/kata-containers-runtime-ARM-18.04-PR/1304/consoleFull for instance...
Yeah, the size of those full logs still bugs me....

**Comment by tedyu** [comment]
```
  0     0    0     0    0     0      0      0 --:--:-- --:--:-- --:--:--     0curl: (7) Failed to connect to 10.110.29.214 port 8080: Connection refused
18:10:04 # service "my-service" deleted
18:10:04 # deployment.apps "hello-world" deleted
18:10:04 Failed at 80: bats "${K8S_TEST_ENTRY}"
...
18:10:04 W1205 02:10:04.227663   11187 removeetcdmember.go:61] [reset] failed to remove etcd member: error syncing endpoints with etc: etcdclient: no available endpoints
```
It seems the failure was not related to my PR.

@jodh-intel 
Mind re-triggering the failed tests?

thanks

**Comment by tedyu** [comment]
Looking at other tests, such as ci-cri-containerd, the test runs passed.

**Comment by tedyu** [comment]
@amshinde @egernst 
Mind triggering another test run ?

thanks

**Comment by tedyu** [comment]
@lifupan 
Can you trigger another test run ?

thanks

**Comment by tedyu** [comment]
@ericooper @egernst 
Please review.

**Comment by tedyu** [comment]
@ericooper @egernst @grahamwhaley 
The condition for returning error currently is backwards.
This means when supported major version goes to 1, the error would be returned for supported version:
```
clh.version.Major < supportedMajorVersion
```
I think we should fix this bug now.

**Comment by tedyu** [comment]
@ericooper @egernst @grahamwhaley
Please take another look.

Thanks

**Comment by amshinde** [comment]
@tedyu You need to gofmt `virtcontainers/clh_test.go` in place using 
```
gofmt -w virtcontainers/clh_test.go
```

This is the diff that needs to be applied to make CI happy :
```
diff --git a/virtcontainers/clh_test.go b/virtcontainers/clh_test.go
index 6e8fb5b7..0cd6f817 100644
--- a/virtcontainers/clh_test.go
+++ b/virtcontainers/clh_test.go
@@ -66,4 +66,4 @@ func TestCheckVersion(t *testing.T) {
                        assert.Error(err, msg)
                }
        }
-}
\ No newline at end of file
+}
```

**Comment by tedyu** [comment]
@amshinde 
I have applied the above diff.
Please take another look.

**Comment by tedyu** [comment]
```
06:39:46 time="2019-12-10T14:39:46Z" level=error msg="error copying docker://quay.io/crio/busybox:latest to dir:/tmp/jenkins/workspace/kata-containers-runtime-ubuntu-18-04-PR/go/src/github.com/cri-o/cri-o/.artifacts/busybox-image: Error initializing source docker://quay.io/crio/busybox:latest: unexpected http code: 504 (Gateway Timeout), URL: https://quay.io/v2/auth?scope=repository%3Acrio%2Fbusybox%3Apull&service=quay.io"
06:39:46 Error pulling quay.io/crio/busybox
```
Not related to my change.

**Comment by tedyu** [comment]
From firecracker test output:
```
Failed at 44: sudo -E PATH="$PATH" bash -c "make crio"
```
Not related to this PR.

**Comment by tedyu** [comment]
ping @kata-containers/runtime for review

File: virtcontainers/clh.go [hunk 3]
```diff
@@ -122,6 +122,18 @@ var clhDebugKernelParams = []Param{
 //
 //###########################################################
 
+func (clh *cloudHypervisor) checkVersion() error {
+	if clh.version.Major > supportedMajorVersion || (clh.version.Major == supportedMajorVersion && clh.version.Minor > supportedMinorVersion) {
```
Comment (bergwolf) [inline virtcontainers/clh.go]: what is the version compatibility policy of clh? Why are we not filtering minimal supported version (as implemented originally in clh.go)?

Comment (jodh-intel) [reply virtcontainers/clh.go]: /cc @sboeuf.

Comment (tedyu) [reply virtcontainers/clh.go]: @bergwolf 
w.r.t. 'filtering minimal supported version', I think you were referring to some code which is not currently in clh.go

Comment (sboeuf) [reply virtcontainers/clh.go]: We don't really have any policy yet regarding the compatibility of different versions. The project just released `0.4.0` but we're not yet at `1.0`.
/cc @sameo

Comment (bergwolf) [reply virtcontainers/clh.go]: @tedyu I was thinking something like:
```
if clh.version.Major > supportedMajorVersion || (clh.version.Major == supportedMajorVersion && clh.version.Minor >= supportedMinorVersion) {
    // proceed
} else {
   // fail
}
```
Then we can define `supportedMajor/MinorVersion` as minimal supported version instead of maximum supported version.

Comment (tedyu) [reply virtcontainers/clh.go]: In that case, currently minimum supportedMajorVersion would be 0.
What value would minimum supportedMinorVersion be (3) ?

Comment (bergwolf) [reply virtcontainers/clh.go]: The code is to implement mechanism. We can bump minimal supported version when needed.

Comment (tedyu) [reply virtcontainers/clh.go]: @bergwolf 
I have addressed your comment.

**Comment by tedyu** [comment]
@bergwolf 
Can you take another look ?

**Comment by jodh-intel** [comment]
Ping @kata-containers/runtime.

/test

**Comment by devimc** [comment]
@tedyu some CIs are failing, please take a look

**Comment by tedyu** [comment]
From http://jenkins.katacontainers.io/job/kata-containers-runtime-cri-containerd-PR/1934/console:
```
--- FAIL: TestContainerListStatsWithSandboxIdFilter (19.39s)
    container_stats_test.go:192: Create a pod config and run sandbox container
    container_stats_test.go:200: Create a container config and run containers in a pod
    container_stats_test.go:222: Fetch container stats for each container with Filter
    container_stats_test.go:239: Verify container stats for sandbox "a54f2a922d04d90bddfa588e9eae0dad3d73561eb93b42e5de830af16e4d82f3"
    require.go:858: 
        	Error Trace:	container_stats_test.go:341
        	            				container_stats_test.go:241
        	Error:      	Should NOT be empty, but was 0
        	Test:       	TestContainerListStatsWithSandboxIdFilter
```
From http://jenkins.katacontainers.io/job/kata-containers-runtime-centos-7-4-q-35-PR/2556/console :
```
Stderr: Error unpacking rpm package gnome-keyring-3.34.0-1.fc31.x86_64
Error: Transaction failed
```
Not related to my PR.

**Comment by raravena80** [comment]
@tedyu ping. Any updates? Thanks.

**Comment by tedyu** [comment]
http://jenkins.katacontainers.io/job/kata-containers-runtime-ARM-18.04-PR/1433/  gave me 404

I checked the test result some time back.
I don't think the test failure was related.

**Comment by tedyu** [comment]
From http://jenkins.katacontainers.io/job/kata-containers-runtime-opensuse-15-PR/1073/console :
```
Stderr: Error unpacking rpm package shadow-utils-2:4.6-9.fc30.x86_64Error: Transaction failed
```

**Comment by fidencio** [comment]
> From http://jenkins.katacontainers.io/job/kata-containers-runtime-opensuse-15-PR/1073/console :
> 
> ```
> Stderr: Error unpacking rpm package shadow-utils-2:4.6-9.fc30.x86_64Error: Transaction failed
> ```

This is a known issue, please refer to: https://github.com/kata-containers/runtime/issues/2580

**Comment by tedyu** [comment]
From the centos test suite:
```
[3] Running command '/usr/bin/docker [docker update --cpus -3.000000 JVebMIP8aULrf5MRY3cdscfKYyhGB3]'
[3] command failed error 'exit status 1'
[3] [docker update --cpus -3.000000 JVebMIP8aULrf5MRY3cdscfKYyhGB3]
[3] Timeout: 120 seconds
[3] Exit Code: 1
```

**Comment by tedyu** [comment]
From ARM ubuntu build log:
```
21:17:13 go: github.com/go-critic/go-critic@v0.0.0-20181204210945-ee9bf5809ead: invalid pseudo-version: does not match version-control timestamp (2019-02-10T22:04:43Z)
21:17:13 Makefile:8: recipe for target 'build' failed
21:17:13 make: *** [build] Error 1
```
Not related to the PR.

@fidencio @raravena80 
Can you take another look ?

**Comment by Pennyzct** [comment]
Hi~ @tedyu ARM failure fixed by https://github.com/kata-containers/tests/pull/2435. And I've rebuilt it. ;)

**Comment by tedyu** [comment]
From power8 build:
```
panic: ERROR: Docker-like container engine not accessible to current user:  (error exec: "podman": executable file not found in $PATH)

goroutine 1 [running, locked to thread]:
```

**Comment by tedyu** [comment]
@Pennyzct @raravena80 
All tests passed

**Comment by tedyu** [comment]
@fidencio @devimc @jodh-intel 
Please review.

**Comment by fidencio** [comment]
@bergwolf, can you confirm all your comments were addressed?

**Comment by amshinde** [comment]
@tedyu Was this change backported to stable?

**Comment by tedyu** [comment]
I have been busy with day job and haven't a chance to do the backport.
